synapseclient/client.py

```diff
@@ -1770,82 +1773,146 @@ def returnDict(destination):
                      'files': [None] if destination is None else [os.path.basename(destination)],
                      'cacheDir': None if destination is None else os.path.dirname(destination) }
 
-        if expected_md5 == '':
-            expected_md5 = None
+        return returnDict(self._download_with_retries(url, destination, file_handle_id, expected_md5))
 
-        ## download via http unless we're redirected to some other scheme
-        download_via_http = True
 
-        # We expect to be redirected to a signed S3 URL or externalURL
-        #The assumption is wrong - we always try to read either the outer or inner requests.get
-        #but sometimes we don't have something to read.  I.e. when the type is ftp at which point
-        #we still set the cache and filepath based on destination which is wrong because nothing was fetched
-        response = _with_retry(lambda: requests.get(url, headers=self._generateSignedHeaders(url), allow_redirects=False), verbose=self.debug, **STANDARD_RETRY_PARAMS)
+    def _download_with_retries(self, url, destination, file_handle_id=None, expected_md5=None, retries=5):
+        while retries > 0:
+            try:
+                return self._download(url, destination, file_handle_id, expected_md5)
+            except SynapseDownloadError as ex:
+                exc_info = sys.exc_info()
+                retries -= 1
+        ## Re-raise exception
+        raise exc_info[0](exc_info[1])
+
+
+    def _download(self, url, destination, file_handle_id=None, expected_md5=None):
+        """
+        Download a file from the given URL to the local file system.
 
-        if response.status_code in [301,302,303,307,308]:
-            url = response.headers['location']
+        :param url: source of download
+        :param destination: destination on local file system
+        :param file_handle_id: (optional) if given, the file will be given a
+                               temporary name that includes the file handle id
+                               which allows resuming partial downloads of the same
+                               file from previous sessions
+        :param expected_md5:   (optional) if given, check that the MD5 of the
```

Comment:
I broke it up for clarity - keeping the retry logic separate from the loop that handles redirects. 

Also, I like that _download_with_retries keeps the original URL in its scope, whereas _downloadFile is free to modify its copy of the URL as it handles redirects. Otherwise, you'd have to stash a copy of the original URL so you could refresh the (time limited) presigned S3 URL when you get a partially completed download.